assets/js/app.js

```diff
@@ -0,0 +1,390 @@
+    };
+  };
+
+  const getCoinHTML = async (coins) => {
+    const selectedCoins =
+      JSON.parse(localStorage.getItem("selectedCoins")) || [];
+
+    const promises = coins.map(async (coin) => {
+      let { id, symbol, name } = coin;
+      let small, usd, eur, ils;
+      try {
+        const coinData = selectedCoins.find((coin) => coin.id === id);
+
+        if (
+          isCoinSelected(id) &&
+          coinData &&
+          getCurrentTime() - coinData.time <= 2
+        ) {
+          ({ small, usd, eur, ils } = coinData);
+        } else {
+          ({ small, usd, eur, ils } = await fetchCoinDetails(id));
+        }
+
+        console.log(coinData);
+        return `
+          <div class="card" id="${id}">
+            <div class="card-body">
+              <div class="switchFlex">
+                <h5 class="card-title">${symbol.toUpperCase()}</h5>
+                <div class="form-check form-switch" >
+                  <input
+                  id="${id}check"
+                    class="form-check-input"
+                    type="checkbox"
+                    role="switch"
+                    ${isCoinSelected(id) ? `checked` : ""}
+                  />
+                </div>
+              </div>
+              <p class="card-text">${name}</p>
+              <button 
+                type="button" 
+                class="btn btn-primary" 
+                data-bs-toggle="popover" 
+                data-bs-title="Coin Info" 
+                data-bs-content="<div style='text-align: center;'><img src='${small}' alt='coin image' style='width: 100px; height: 100px;' /><br/>USD Price: ${usd}$ <br/>EUR Price ${eur}€ <br/> ILS Price ${ils}₪</div>">
+                More Details
+              </button>
+            </div>
+          </div>
+        `;
+      } catch (error) {
+        console.error(`Error fetching details for ${id}:`, error);
+        return `
+          <div class="card" id="${id}">
+            <div class="card-body">
+              <div class="switchFlex">
+                <h5 class="card-title">${symbol.toUpperCase()}</h5>
+                <div class="form-check form-switch >
+                  <input
+                  id="${id}check"
+                    class="form-check-input"
+                    type="checkbox"
+                    role="switch"
+                    ${isCoinSelected(id) ? `checked` : ""}
+                  />
+                </div>
+              </div>
+              <p class="card-text">${name}</p>
+              <p class="text-danger">Failed to load additional details</p>
+            </div>
+          </div>
+        `;
+      }
+    });
+
+    const htmlArray = await Promise.all(promises);
+    return htmlArray.join("");
+  };
+
+  const renderAllCoins = (newHTML) => {
+    document.getElementById("coinMain").innerHTML = newHTML;
+    activatePopovers();
+    setupToggleButtons(newHTML);
+  };
+
+  const activatePopovers = () => {
```

Comment:
the popovers act weird. they should close if they lose focus, or at least when i open another popover (i can fill the page with open popovers...)